modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PartitionMeta.java

```diff
@@ -46,212 +45,33 @@ public class PartitionMeta {
-     */
-    public long lastAppliedTerm() {
-        return lastAppliedTerm;
-    }
-
-    /**
-     * Sets last applied index and term.
-     *
-     * @param checkpointId Checkpoint ID.
-     * @param lastAppliedIndex Last applied index value.
-     * @param lastAppliedTerm Last applied term value.
-     */
-    public void lastApplied(@Nullable UUID checkpointId, long lastAppliedIndex, long lastAppliedTerm) {
-        updateSnapshot(checkpointId);
-
-        this.lastAppliedIndex = lastAppliedIndex;
-        this.lastAppliedTerm = lastAppliedTerm;
-    }
-
-    /**
-     * Returns ID of the first page in a chain storing a blob representing last replication protocol group config.
-     */
-    public long lastReplicationProtocolGroupConfigFirstPageId() {
-        return lastReplicationProtocolGroupConfigFirstPageId;
-    }
-
-    /**
-     * Sets ID of the first page in a chain storing a blob representing last replication protocol group config.
-     *
-     * @param checkpointId Checkpoint ID.
-     * @param pageId PageId.
-     */
-    public void lastReplicationProtocolGroupConfigFirstPageId(@Nullable UUID checkpointId, long pageId) {
-        updateSnapshot(checkpointId);
-
-        this.lastReplicationProtocolGroupConfigFirstPageId = pageId;
-    }
-
-    /**
-     * Returns version chain tree root page ID.
-     */
-    public long versionChainTreeRootPageId() {
-        return versionChainTreeRootPageId;
-    }
-
-    /**
-     * Sets version chain root page ID.
-     *
-     * @param checkpointId Checkpoint ID.
-     * @param versionChainTreeRootPageId Version chain root page ID.
-     */
-    public void versionChainTreeRootPageId(@Nullable UUID checkpointId, long versionChainTreeRootPageId) {
-        updateSnapshot(checkpointId);
-
-        this.versionChainTreeRootPageId = versionChainTreeRootPageId;
-    }
-
-    /**
-     * Returns free list root page ID.
-     */
-    public long freeListRootPageId() {
-        return freeListRootPageId;
-    }
-
-    /**
-     * Sets free list root page ID.
-     *
-     * @param checkpointId Checkpoint ID.
-     * @param freeListRootPageId Free list root page ID.
-     */
-    public void freeListRootPageId(@Nullable UUID checkpointId, long freeListRootPageId) {
-        updateSnapshot(checkpointId);
-
-        this.freeListRootPageId = freeListRootPageId;
-    }
-
-    /**
-     * Returns index meta tree meta page id.
-     */
-    public long indexTreeMetaPageId() {
-        return indexTreeMetaPageId;
-    }
-
-    /**
-     * Sets an index meta tree meta page id.
-     *
-     * @param checkpointId Checkpoint id.
-     * @param indexTreeMetaPageId Index meta tree meta page id.
+     * @param pageCount Page count.
      */
-    public void indexTreeMetaPageId(@Nullable UUID checkpointId, long indexTreeMetaPageId) {
-        updateSnapshot(checkpointId);
+    protected PartitionMeta(@Nullable UUID checkpointId, int pageCount) {
+        this(checkpointId);
```

Comment:
Is this a safe constructor call?
It calls `buildSnapshot(checkpointId);` inside of it, which is an abstract method, but `pageCount` is not yet initialized. IDEA should show warnings in such cases. I think it's broken and you should explicitly initialize all fields here and explicitly call `buildSnapshot(checkpointId);` instead of delegating work to another constructor.

Reply:
You are right, it was broken. I think, a lot of confusion occurred because objects of different types were passed to each other's constructors. MetaIo was used in Meta's constructor, and Snapshot was created by passing Meta object.
I removed that and made "plain" constructor, that receives just required fields.
@rpuch , @ibessonov could you please recheck it? It became a bit more verbose, but it seem to me more bug-safe and readable